### ISSUE TITLE
如何做密态下的index？

### ISSUE THREAD

**Issue by nantsing** [secretflow/secretflow#380]
### Issue Type

Bug

### Source

binary

### Secretflow Version

secretflow-0.7.18b4

### OS Platform and Distribution

Linux centos 3.10.0

### Python version

3.8.13

### What happend and What you expected to happen.

```shell
想要在密态下做索引，如运行下述代码，但会报错说索引必须是public的，请问这种情况有解决方案吗？
```

### Reproduction code to reproduce the issue.

```shell
def func(x, y):
    idx = jnp.argmax(x)
    return y[idx]

def alice_get_data():
    return jnp.array([1,2,3,4])

def bob_get_data():
    return jnp.array([1,2,3,4])

if __name__ == '__main__':
    sf.shutdown()
    sf.init(['alice', 'bob'], address='local', log_to_driver=True)
    alice, bob = sf.PYU('alice'), sf.PYU('bob')
    spu = sf.SPU(sf.utils.testing.cluster_def(['alice', 'bob']))
    device = spu
    x, y = alice(alice_get_data)(), bob(bob_get_data)()
    result = device(func)(x, y)
    print(sf.reveal(result))


报错：
  File "/home/jiahaozhao/.conda/envs/secretflow/lib/python3.8/site-packages/secretflow/device/device/spu.py", line 342, in run
    self.runtime.run(executable)
  File "/home/jiahaozhao/.conda/envs/secretflow/lib/python3.8/site-packages/spu/binding/api.py", line 46, in run
    return self._vm.Run(executable.SerializeToString())
RuntimeError: what: 
        [Enforce fail at spu/kernel/hlo/utils.cc:31] value.isPublic(). indicies value must be public.
```

**Comment by tpppppub** [comment]
你好，SPU 目前暂不支持 secret indexing， 请参考 [issue](https://github.com/secretflow/spu/issues/116#issuecomment-1463731396)

**Comment by nantsing** [comment]
> 你好，SPU 目前暂不支持 secret indexing， 请参考 [issue](https://github.com/secretflow/spu/issues/116#issuecomment-1463731396)

好的，感谢解答

**Comment by icavan** [comment]
> > 你好，SPU 目前暂不支持 secret indexing， 请参考 [issue](https://github.com/secretflow/spu/issues/116#issuecomment-1463731396)
> 
> 好的，感谢解答

Secret indexing requires O(N) computation and communication effort, which will be quite slow compared to the plaintext world.

We've some plans to make it available in future releases.

Could you show the cases why you need the secret indexing?

**Comment by nantsing** [comment]
> > > 你好，SPU 目前暂不支持 secret indexing， 请参考 [issue](https://github.com/secretflow/spu/issues/116#issuecomment-1463731396)
> > 
> > 
> > 好的，感谢解答
> 
> Secret indexing requires O(N) computation and communication effort, which will be quite slow compared to the plaintext world.
> 
> We've some plans to make it available in future releases.
> 
> Could you show the cases why you need the secret indexing?

我想用argmax找到x数组最大值对应的下标，然后用这个下标去找y数组里对应的值，但argmax返回的index是被加密的

**Comment by icavan** [comment]
> > > > 你好，SPU 目前暂不支持 secret indexing， 请参考 [issue](https://github.com/secretflow/spu/issues/116#issuecomment-1463731396)
> > > 
> > > 
> > > 好的，感谢解答
> > 
> > 
> > Secret indexing requires O(N) computation and communication effort, which will be quite slow compared to the plaintext world.
> > We've some plans to make it available in future releases.
> > Could you show the cases why you need the secret indexing?
> 
> 我想用argmax找到x数组最大值对应的下标，然后用这个下标去找y数组里对应的值，但argmax返回的index是被加密的

The secret ArgMax result, say secret index 2 could be converted into a secret one hot array [0, 0, 1, 0, 0]. Then a final dot product will give you the secret indexing result.

**Comment by nantsing** [comment]
> > > > > 你好，SPU 目前暂不支持 secret indexing， 请参考 [issue](https://github.com/secretflow/spu/issues/116#issuecomment-1463731396)
> > > > 
> > > > 
> > > > 好的，感谢解答
> > > 
> > > 
> > > Secret indexing requires O(N) computation and communication effort, which will be quite slow compared to the plaintext world.
> > > We've some plans to make it available in future releases.
> > > Could you show the cases why you need the secret indexing?
> > 
> > 
> > 我想用argmax找到x数组最大值对应的下标，然后用这个下标去找y数组里对应的值，但argmax返回的index是被加密的
> 
> The secret ArgMax result, say secret index 2 could be converted into a secret one hot array [0, 0, 1, 0, 0]. Then a final dot product will give you the secret indexing result.

请问一下这个密态的one hot具体怎么转换啊，我用jnp.eye好像还是有点问题，报错是一样的

**Comment by icavan** [comment]
> > > > > > 你好，SPU 目前暂不支持 secret indexing， 请参考 [issue](https://github.com/secretflow/spu/issues/116#issuecomment-1463731396)
> > > > > 
> > > > > 
> > > > > 好的，感谢解答
> > > > 
> > > > 
> > > > Secret indexing requires O(N) computation and communication effort, which will be quite slow compared to the plaintext world.
> > > > We've some plans to make it available in future releases.
> > > > Could you show the cases why you need the secret indexing?
> > > 
> > > 
> > > 我想用argmax找到x数组最大值对应的下标，然后用这个下标去找y数组里对应的值，但argmax返回的index是被加密的
> > 
> > 
> > The secret ArgMax result, say secret index 2 could be converted into a secret one hot array [0, 0, 1, 0, 0]. Then a final dot product will give you the secret indexing result.
> 
> 请问一下这个密态的one hot具体怎么转换啊，我用jnp.eye好像还是有点问题，报错是一样的



@anakinxc  please take a look at `np.eye` support, if this interface makes sense, please add it to the future support list

--

@nantsing  One way is to create a public array `pub_array = [0, 1, 2, 3, 4]`. Then create a secret array `index_array [2, 2, 2, 2, 2]`. Compare these two arrays to get the one-hot array, the comparison results will be either one or zero.

**Comment by nantsing** [comment]
> > > > > > > 你好，SPU 目前暂不支持 secret indexing， 请参考 [issue](https://github.com/secretflow/spu/issues/116#issuecomment-1463731396)
> > > > > > 
> > > > > > 
> > > > > > 好的，感谢解答
> > > > > 
> > > > > 
> > > > > Secret indexing requires O(N) computation and communication effort, which will be quite slow compared to the plaintext world.
> > > > > We've some plans to make it available in future releases.
> > > > > Could you show the cases why you need the secret indexing?
> > > > 
> > > > 
> > > > 我想用argmax找到x数组最大值对应的下标，然后用这个下标去找y数组里对应的值，但argmax返回的index是被加密的
> > > 
> > > 
> > > The secret ArgMax result, say secret index 2 could be converted into a secret one hot array [0, 0, 1, 0, 0]. Then a final dot product will give you the secret indexing result.
> > 
> > 
> > 请问一下这个密态的one hot具体怎么转换啊，我用jnp.eye好像还是有点问题，报错是一样的
> 
> @anakinxc please take a look at `np.eye` support, if this interface makes sense, please add it to the future support list
> 
> --
> 
> @nantsing One way is to create a public array `pub_array = [0, 1, 2, 3, 4]`. Then create a secret array `index_array [2, 2, 2, 2, 2]`. Compare these two arrays to get the one-hot array, the comparison results will be either one or zero.

OK, thanks!

**Comment by anakinxc** [comment]
Fixed via #395
